### ISSUE TITLE
Port Haar/DCT time reparameterization (time_reparam) from Pyro

### ISSUE THREAD

**Issue by juanitorduz** [juanitorduz/numpyro_forecast#12]
## Summary

Pyro's `pyro.contrib.forecast` supports **time-axis reparameterization** of all time-dependent latent sites via the `time_reparam` argument (`"haar"` for the Haar wavelet transform, `"dct"` for the discrete cosine transform). This is one of Pyro's most useful features for forecasting: it decorrelates random-walk-style latents across time and dramatically improves SVI/NUTS conditioning and convergence.

`numpyro_forecast` does **not** port this feature, because the underlying primitives do not exist in NumPyro.

## Why it's currently not portable

In Pyro the feature is wired up with a single line in `Forecaster` / `HMCForecaster`:

```python
if time_reparam == "haar":
    model = poutine.reparam(model, time_reparam_haar)
elif time_reparam == "dct":
    model = poutine.reparam(model, time_reparam_dct)
```

and the config functions (in `pyro/contrib/forecast/util.py`) inspect the `time` plate frame and return a `HaarReparam` / `DiscreteCosineReparam` for every non-observed site inside it:

```python
def time_reparam_haar(msg):
    if msg["is_observed"]:
        return
    for frame in msg["cond_indep_stack"]:
        if frame.name == "time":
            dim = frame.dim - msg["fn"].event_dim
            return HaarReparam(dim=dim, experimental_allow_batch=True)
```

`HaarReparam` and `DiscreteCosineReparam` are imported from `pyro.infer.reparam`.

**The blockers in NumPyro:**

- `numpyro.infer.reparam` ships only `LocScaleReparam`, `TransformReparam`, `NeuTraReparam`, `CircularReparam`, `ProjectedNormalReparam`, and `ExplicitReparam`. There is **no `HaarReparam` and no `DiscreteCosineReparam`**.
- `numpyro.distributions.transforms` has **no Haar or DCT transform** either.

So the entire `time_reparam` mechanism rests on two transform/reparam classes that simply don't exist in the JAX/NumPyro ecosystem.

**What *does* already work:** `numpyro.handlers.reparam` accepts a *callable* `config` (`config(msg) -> Reparam | None`), so Pyro's plate-frame-inspection wiring translates almost verbatim. The only missing pieces are the two reparam classes and their underlying transforms.

## Potential solution: port the primitives from Pyro

Port the transform + reparam stack to JAX/NumPyro, then wire it into the forecasters.

1. **Transforms** (new, in a module like `numpyro_forecast/transforms.py` or upstreamed to NumPyro):
   - `HaarTransform` — orthonormal Haar wavelet transform along a chosen axis (Pyro: `pyro.ops.tensor_utils.haar_transform` / `pyro.distributions.transforms.HaarTransform`). Implement `forward`, `inverse`, and a constant `log_abs_det_jacobian` (orthonormal ⇒ 0). Needs to handle non-power-of-2 lengths the way Pyro does.
   - `DiscreteCosineTransform` — orthonormal DCT-II/III along a chosen axis (Pyro: `pyro.ops.tensor_utils.dct`/`idct`, `pyro.distributions.transforms.DiscreteCosineTransform`). JAX has `jax.scipy.fft.dct`/`idct` which can back this, keeping it differentiable.
2. **Reparam classes** mirroring Pyro's `HaarReparam` / `DiscreteCosineReparam` (subclasses of `numpyro.infer.reparam.Reparam`), each applying the corresponding transform along the time axis, including the `experimental_allow_batch` semantics (reshape so the transformed axis is the rightmost event dim).
3. **Wiring** in `Forecaster` / `HMCForecaster` (and `fit_svi` / `fit_mcmc`): add a `time_reparam: Literal["haar", "dct"] | None = None` argument that wraps the model with `numpyro.handlers.reparam` using a callable config that targets sites inside the `time` plate, mirroring Pyro's `time_reparam_haar` / `time_reparam_dct`.
4. **Tests**: round-trip transform correctness (forward∘inverse ≈ identity, Jacobian ≈ 0), and an end-to-end SVI/NUTS forecasting test showing the reparam runs and produces finite forecasts.

## References

- Pyro forecaster: https://github.com/pyro-ppl/pyro/blob/dev/pyro/contrib/forecast/forecaster.py
- Pyro reparam config functions: https://github.com/pyro-ppl/pyro/blob/dev/pyro/contrib/forecast/util.py
- Pyro reparam classes: `pyro.infer.reparam.HaarReparam`, `pyro.infer.reparam.DiscreteCosineReparam`
- Pyro transforms: `pyro.distributions.transforms.HaarTransform`, `pyro.distributions.transforms.DiscreteCosineTransform`

## Notes / open questions

- Exact `"haar"`/`"dct"` → reparam-class mapping should be confirmed against current Pyro source when implementing.
- Worth deciding whether these transforms belong in this package or should be contributed upstream to NumPyro (more broadly useful).
- This was deliberately left out of the initial release as out of scope; this issue tracks it as a follow-up enhancement.
